### ISSUE TITLE
openssl version used is insecure

### ISSUE THREAD

**Issue by vvitayau** [docker/compose#1601]
> $ docker-compose --version
> docker-compose version: 1.3.1
> CPython version: 2.7.9
> OpenSSL version: OpenSSL 1.0.1j 15 Oct 2014

see https://www.openssl.org/news/secadv_20150319.txt that mentions

```

Reclassified: RSA silently downgrades to EXPORT_RSA [Client] (CVE-2015-0204)
============================================================================

Severity: High

This security issue was previously announced by the OpenSSL project and
classified as "low" severity. This severity rating has now been changed to
"high".

This was classified low because it was originally thought that server RSA
export ciphersuite support was rare: a client was only vulnerable to a MITM
attack against a server which supports an RSA export ciphersuite. Recent
studies have shown that RSA export ciphersuites support is far more common.

This issue affects OpenSSL versions: 1.0.1, 1.0.0 and 0.9.8.

OpenSSL 1.0.1 users should upgrade to 1.0.1k.
OpenSSL 1.0.0 users should upgrade to 1.0.0p.
OpenSSL 0.9.8 users should upgrade to 0.9.8zd.

This issue was reported to OpenSSL on 22nd October 2014 by Karthikeyan
Bhargavan of the PROSECCO team at INRIA. The fix was developed by Stephen
Henson of the OpenSSL core team. It was previously announced in the OpenSSL
security advisory on 8th January 2015.
```

**Comment by vvitayau** [comment]
https://www.openssl.org/news/openssl-1.0.1-notes.html mentions __OpenSSL 1.0.1o [12 Jun 2015]__
but current latest 1.0.1l is available in https://github.com/Homebrew/homebrew/commits/master/Library/Formula/openssl.rb

**Comment by aanand** [comment]
Unfortunately, 1.0.2 has an incompatibility with Python's `ssl` module, causing certificate verification errors - see #1484.

Long-term, we should use pyOpenSSL to avoid this; for now, we're stuck with 1.0.1. We might have to build 1.0.1o ourselves in the Dockerfile (for Linux) and `script/prepare-osx` (for OSX).

**Comment by aanand** [comment]
#1530 is the pyOpenSSL tracking issue.

**Comment by unserializable** [comment]
I have freshly downloaded docker-compose 1.5.1 binary for Linux-x86_64 and what do I see from 'docker-compose version' output -- OpenSSL 1.0.1e from 11 Feb 2013 -- even older than one reported by @vvitayau :

docker-compose version: 1.5.1
docker-py version: 1.5.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013

**Comment by aanand** [comment]
Update on this: the root cause of the problems we were having was actually a boot2docker issue. Using the OpenSSL 1.0.1 branch was a workaround, as explained in https://github.com/docker/compose/issues/890#issuecomment-127490952:

> The Subject and Issuer DNs of the server certificate generated by boot2docker are identically set to `/O=Boot2Docker`. I believe the server certificate is actually signed by the CA cert, but AFAICT OpenSSL 1.0.2 uses this information (i.e. identical Subject and Issuer DNs) to reject the server certificate as self-signed instead of attempting to verify the server certificate against the provided CA cert. Versions of OpenSSL prior to 1.0.2 validate the server certificate against the provided CA cert, which succeeds.

The boot2docker issue was fixed in https://github.com/boot2docker/boot2docker/pull/1029 and made it into boot2docker 1.8. The current version of Compose requires a minimum Engine version of 1.7. If we update that minimum to 1.8 (or higher) with the next Compose release, we should be able to use the latest OpenSSL without worries.

In general, when building binaries we should probably have the script ensure we're using as up-to-date a version of OpenSSL as possible.

**Comment by thaJeztah** [comment]
Should this be on the next milestone?

**Comment by vvitayau** [comment]
I've noticed that ``docker-compose --version``  no longer mention which openssl version

And is this issue no longer relevant if using https://blog.docker.com/2016/03/docker-for-mac-windows-beta ?

Should I now instead rely on Apple version of ``/usr/bin/openssl`` 
...     "Apple has deprecated use of OpenSSL in favor of its own TLS and crypto libraries"

In the past, I've been using https://github.com/Homebrew/homebrew-versions/blob/master/openssl101.rb
> 
$ ln -snf /usr/local/Cellar/openssl101/1.0.1r/bin/openssl /usr/local/bin/openssl
$ which openssl
/usr/local/bin/openssl
$ openssl version
OpenSSL 1.0.1r  28 Jan 2016

With
> 
$ docker-compose --version
docker-compose version 1.6.2, build 4d72027
$ docker --version
Docker version 1.10.3, build 20f81dd

**Comment by aanand** [comment]
> I've noticed that `docker-compose --version` no longer mention which openssl version

Use `docker-compose version`, not `docker-compose --version`, to get extended version information.

> Is this issue no longer relevant if using https://blog.docker.com/2016/03/docker-for-mac-windows-beta ???

It's still relevant - the Docker for Mac beta replaces Virtualbox and Docker Machine, not Compose.

**Comment by BradleyA** [comment]
When is security going to become more important than a new insecure feature?  This should be using  OpenSSL 1.0.2j  26 Sep 2016 or at least something in 2016 not 2013!

ubuntu $ docker-compose version
docker-compose version 1.8.1, build 878cff1
docker-py version: 1.10.3
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013

ubuntu $ openssl version
OpenSSL 1.0.2j  26 Sep 2016

**Comment by shin-** [comment]
Closing - Was fixed last year: #3489; #4009
